filebeat/input/filestream/active_monitor.go

```diff
@@ -0,0 +1,100 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package filestream
+
+import (
+	"sync"
+	"time"
+
+	loginp "github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/go-concert/unison"
+)
+
+// activeFileMonitor checks the state of every opened file.
+// If a file is renamed or removed, it cancels the harvester.
+// This has to run separately from reading the file
+// to avoid blocking when the output cannot accept events.
+type activeFileMonitor struct {
+	sync.RWMutex
+	log   *logp.Logger
+	files map[string]fileSource
+
+	interval     time.Duration
+	closeRenamed bool
+	closeRemoved bool
+}
+
+func newActiveFileMonitor(cfg stateChangeCloserConfig) *activeFileMonitor {
+	return &activeFileMonitor{
+		log:          logp.NewLogger("active_file_monitor"),
+		files:        make(map[string]fileSource, 0),
+		interval:     cfg.CheckInterval,
+		closeRenamed: cfg.Renamed,
+		closeRemoved: cfg.Removed,
+	}
+}
+
+func (m *activeFileMonitor) addFile(path string, src fileSource) bool {
+	m.Lock()
+	defer m.Unlock()
+
+	m.log.Debug("Adding new file to monitor %s", path)
+	if _, ok := m.files[path]; ok {
+		return false
+	}
+	m.files[path] = src
+	return true
+}
+
+func (m *activeFileMonitor) run(ctx unison.Canceler, hg *loginp.HarvesterGroup) {
+	ticker := time.NewTicker(m.interval)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			m.Lock()
+			for path, src := range m.files {
+				isSame, err := isSameFile(path, src.info)
+				if m.closeBecauseRemoved(path, err) || m.closeBecauseRenamed(path, isSame) {
```

Comment:
I'm trying to understand `closeBecauseRenamed`... elsewhere, the way the filestream input detects if something is renamed is if `os.SameFile` returns true for a different path than the original, indicating that the file has been moved to that new path. Here though, it seems to instead check whether `os.SameFile` returns _false_ for the _same_ path as before. To me all that indicates is that something else is where that file used to be, and isn't enough to conclude whether the original was removed or renamed.

Here's the specific case that concerns me: if `closeRemoved == false` and `closeRenamed == true`, and we remove a file, then `isSameFile` will return `false` with a non-nil error because the `os.Stat` call fails. Then `closeBecauseRemoved` will return false, because `closeRemoved` is false, but `closeBecauseRenamed` will return true even though the file wasn't renamed.

It's possible I'm mis-parsing this, but in that case some clarification in the language or comments here would be appreciated :-)

Reply:
Indeed. :facepalm: :facepalm: :facepalm:  I will move these checks back to the filereader. Thus, I removed it from the PR.